### PR TITLE
[GOBBLIN-423] Add record count limit to salesforce source

### PR DESCRIPTION
**Pull request by yukuai518**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-423

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
   Set fixed upper bound in bootstrapping mode. Don't move the upper bound in the future run.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by yukuai518** [comment]
@zxcware @htran1 Please review

File: gobblin-core/src/main/java/org/apache/gobblin/source/extractor/partition/Partitioner.java [hunk 1]
```diff
@@ -322,6 +333,17 @@ protected long getLowWatermark(ExtractType extractType, WatermarkType watermarkT
         lowWatermark = this.getAppendLowWatermark(watermarkType, previousWatermark, deltaForNextWatermark);
       }
     }
+
+    if (isBootStrapWithFixedHighWatermark()) {
```
Comment (htran1) [inline gobblin-core/src/main/java/org/apache/gobblin/source/extractor/partition/Partitioner.java]: This looks the same as the else block above, so it can be merged by changing the if to ` if (!isBootStrapWithFixedHighWatermark() && (this.isFullDump() || this.isWatermarkOverride())) {`

File: gobblin-core/src/main/java/org/apache/gobblin/source/extractor/partition/Partitioner.java [hunk 2]
```diff
@@ -407,9 +429,30 @@ protected long getHighWatermark(ExtractType extractType, WatermarkType watermark
...
     }
+
+    // Use fixed high watermark in bootstrapping mode
+    if (isBootStrapWithFixedHighWatermark()) {
+      hasUserSpecifiedHighWatermark = true;
+      if (this.state.contains(ConfigurationKeys.SOURCE_QUERYBASED_END_VALUE)) {
+        highWatermark = this.state.getPropAsLong(ConfigurationKeys.SOURCE_QUERYBASED_END_VALUE);
+        LOG.info("Bootstrapping mode: persisted high watermark " + highWatermark);
+      } else {
+        this.state.setProp(ConfigurationKeys.SOURCE_QUERYBASED_END_VALUE, highWatermark);
```
Comment (htran1) [inline gobblin-core/src/main/java/org/apache/gobblin/source/extractor/partition/Partitioner.java]: Is this state getting persisted and propagated to the next run? If it does then what happens when user changes the configuration? Will the config change override the state?

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java [hunk 3]
```diff
@@ -79,6 +79,11 @@
   public static final boolean DEFAULT_USE_ALL_OBJECTS = false;
 
   private static final String ENABLE_DYNAMIC_PARTITIONING = "salesforce.enableDynamicPartitioning";
+  private static final String DYNAMIC_PROBING_TOTAL_RECORDS_LIMIT = "salesforce.dynamicProbing.total.records.limit";
```
Comment (htran1) [inline gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: Camel case names should be used here.

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java [hunk 4]
```diff
@@ -131,7 +136,19 @@
...
 
-    String specifiedPartitions = generateSpecifiedPartitions(histogram, minTargetPartitionSize, maxPartitions,
+    // we should look if the count is too big, cut off early if count exceeds the limit, or bucket size is too large
+    Histogram histogramActual = new Histogram();
+    for (HistogramGroup group: histogram.getGroups()) {
+      if (histogramActual.getTotalRecordCount() + group.count > state.getPropAsLong(DYNAMIC_PROBING_TOTAL_RECORDS_LIMIT, DYNAMIC_PROBING_DEFAULT_TOTAL_RECORDS_LIMIT)) {
+        break;
+      }
+      histogramActual.add(group);
+      if (histogramActual.getGroups().size() >= state.getPropAsLong(DYNAMIC_PROBING_TOTAL_BUCKETS_LIMIT, DYNAMIC_PROBING_DEFAULT_TOTAL_BUCKETS_LIMIT)) {
```
Comment (htran1) [inline gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: Why have the bucket limit here?

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java [hunk 5]
```diff
@@ -131,7 +136,19 @@
     Partition partition = new Partitioner(state).getGlobalPartition(previousWatermark);
     Histogram histogram = getHistogram(sourceEntity.getSourceEntityName(), watermarkColumn, state, partition);
 
-    String specifiedPartitions = generateSpecifiedPartitions(histogram, minTargetPartitionSize, maxPartitions,
+    // we should look if the count is too big, cut off early if count exceeds the limit, or bucket size is too large
+    Histogram histogramActual = new Histogram();
```
Comment (htran1) [inline gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: We should consider pushing this logic into getHistogram() for early termination of the search or add a note here indicating possible future work.

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java (hunk 4 above)
Comment (yukuai518) [reply gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: @htran1 do we have total query limits? I think each bucket is a single query, if we have total limits for querying count, we can set this bucket number.

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java (hunk 3 above)
Comment (yukuai518) [reply gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: @htran1 do you mean you want to change other attributes as well? this is constant, why need Camel case?

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java (hunk 4 above)
Comment (htran1) [reply gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: The query limit is enforced by `salesforce.dynamicProbingLimit`. The loop here doesn't execute any queries, so the limit is not required here.

File: gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java [hunk 6]
```diff
@@ -504,6 +534,14 @@ void add(Histogram histogram) {
       totalRecordCount += histogram.totalRecordCount;
     }
 
+    HistogramGroup get(int idx) {
```
Comment (htran1) [inline gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: Why is this required? I don't see anywhere where the null result is checked.

Comment (yukuai518) [reply gobblin-salesforce/src/main/java/org/apache/gobblin/salesforce/SalesforceSource.java]: I use this to get the latest group.
